src/NuGet.Core/NuGet.Packaging/Signing/Archive/ReadOnlyBufferedStream.cs

```diff
@@ -0,0 +1,365 @@
+            }
+            else if (IsPositionAfterEndOfStream())
+            {
+                return -1;
+            }
+            else
+            {
+                FillBuffer();
+
+                offset = 0;
+            }
+
+            var value = _buffer[offset];
+
+            ++Position;
+
+            return value;
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            ThrowIfDisposed();
+
+            long newOffset;
+
+            switch (origin)
+            {
+                case SeekOrigin.Current:
+                    newOffset = Position + offset;
+                    break;
+
+                case SeekOrigin.Begin:
+                    newOffset = offset;
+                    break;
+
+                case SeekOrigin.End:
+                    newOffset = Length + offset;
+                    break;
+
+                default:
+                    throw new ArgumentException();
+            }
+
+            if (newOffset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            Position = _stream.Seek(newOffset, SeekOrigin.Begin);
+
+            return Position;
+        }
+
+        public override void SetLength(long value)
+        {
+            ThrowIfDisposed();
+
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            ThrowIfDisposed();
+
+            throw new NotSupportedException();
+        }
+
+        public override void WriteByte(byte value)
+        {
+            ThrowIfDisposed();
+
+            throw new NotSupportedException();
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (!_isDisposed)
+                {
+                    if (!_leaveOpen)
+                    {
+                        _stream.Dispose();
+                    }
+
+                    _isDisposed = true;
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        private void FillBuffer()
+        {
+            _bufferStartPosition = Position;
+            _bufferFillLength = 0;
+
+            _stream.Position = _bufferStartPosition;
+
+            var totalBytesRead = 0;
+            var bytesRead = 0;
+            var offset = 0;
+            var count = _buffer.Length;
+
+            do
+            {
+                bytesRead = _stream.Read(_buffer, offset, count);
+
+                offset += bytesRead;
+                count -= bytesRead;
+                totalBytesRead += bytesRead;
+            } while (bytesRead > 0 && count > 0);
```

Comment:
Why do you need this do/while? Can you please add a comment explaining it?

Reply:
`Read(...)` does not guarantee that the requested number of bytes will be read, even if there are ample bytes in the source.  You have to assume that Read(...)` could perform partial read.  The relevant documentation is [here](https://msdn.microsoft.com/en-us/library/system.io.stream.read(v=vs.110).aspx#Anchor_2):

>An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached